### PR TITLE
MAINT: minimize jac docstring closes #9042

### LINKED ISSUE #9042: optimize: Incorrect statement about `jac` in the `minimize` docstring.
In the description of the `jac` argument, the docstring for `scipy.optimize.minimize` says

    Alternatively, the keywords {'2-point', '3-point', 'cs'} select a finite
    difference scheme for numerical estimation of the gradient.

and

    If False, the gradient will be estimated using '2-point' finite difference estimation.

Those descriptions are not correct when `method` is `'Newton-CG'` or `trust-ncg` (and maybe others--I haven't checked).  For example,

```
In [156]: def func(x):
     ...:     return ((x-1)**2).sum()
     ...: 

In [157]: from scipy.optimize import minimize
```

Try `'Newton-CG'` with `jac=False`:

```
In [158]: minimize(func, [1.1, 1.2], method='Newton-CG', jac=False)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-158-94c7580e9508> in <module>()
----> 1 minimize(func, [1.1, 1.2], method='Newton-CG', jac=False)

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    600     elif meth == 'newton-cg':
    601         return _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback,
--> 602                                   **options)
    603     elif meth == 'l-bfgs-b':
    604         return _minimize_lbfgsb(fun, x0, args, jac, bounds,

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/optimize.py in _minimize_newtoncg(fun, x0, args, jac, hess, hessp, callback, xtol, eps, maxiter, disp, return_all, **unknown_options)
   1494     _check_unknown_options(unknown_options)
   1495     if jac is None:
-> 1496         raise ValueError('Jacobian is required for Newton-CG method')
   1497     f = fun
   1498     fprime = jac

ValueError: Jacobian is required for Newton-CG method
```

Try `'trust-ncg'` with `jac=False`:

```
In [159]: minimize(func, [1.1, 1.2], method='trust-ncg', jac=False)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-159-0e8fe542e188> in <module>()
----> 1 minimize(func, [1.1, 1.2], method='trust-ncg', jac=False)

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_minimize.py in minimize(fun, x0, args, method, jac, hess, hessp, bounds, constraints, tol, callback, options)
    621     elif meth == 'trust-ncg':
    622         return _minimize_trust_ncg(fun, x0, args, jac, hess, hessp,
--> 623                                    callback=callback, **options)
    624     elif meth == 'trust-krylov':
    625         return _minimize_trust_krylov(fun, x0, args, jac, hess, hessp,

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.2.0.dev0+6d455f3-py3.5-macosx-10.6-x86_64.egg/scipy/optimize/_trustregion_ncg.py in _minimize_trust_ncg(fun, x0, args, jac, hess, hessp, **trust_region_options)
     32     """
     33     if jac is None:
---> 34         raise ValueError('Jacobian is required for Newton-CG trust-region '
     35                          'minimization')
     36     if hess is None and hessp is None:

ValueError: Jacobian is required for Newton-CG trust-region minimization
```

Apparently a Jacobian *function* is required for these methods.  The `minimize` docstring should be updated to make this clear.  Or, if it is feasible, the methods should be fixed to behave as documented.

If users happen to follow the references to the additional documentation about the [`'Newton-CG` method](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-newtoncg.html#optimize-minimize-newtoncg), they'll see the note that says "Note that the jac parameter (Jacobian) is required."  That's too well hidden, and still not explicit.  Does it mean providing the argument `jac='3-point'` is sufficient?  If an implementation of the Jacobian function (i.e. a callable) is required, it should say so explicitly, and that information should be in the `minimize` docstring.

See https://stackoverflow.com/questions/33926357/jacobian-is-required-for-newton-cg-method-when-doing-a-approximation-to-a-jaco and https://stackoverflow.com/questions/51369984/jacobian-is-required-for-newton-cg-method-issue-still-exists-for-trust-ncg for cases where the current documentation leads to confusion.

### PR DESCRIPTION
**Pull request by andyfaff**
Closes #9042

### CODE REVIEW

**Comment by mdhaber** [comment]
I tested the methods:
```
import numpy as np
from scipy.optimize import minimize

def f(x):
    return np.sum(x**2)

def j(x):
    return 2*x

methods = ["COBYLA", 'Powell', 'nelder-mead',
           'L-BFGS-B', 'BFGS', 'CG', 'TNC', 'SLSQP',
           "Newton-CG", 'dogleg', 'trust-ncg', 'trust-exact',
           'trust-krylov', 'trust-constr']

jacobians = [None, j, '2-point']  # ['three-point', 'cs']
for method in methods:
    for jacobian in jacobians:
        try:
            res = minimize(f, np.random.rand(10), method=method, jac=jacobian)
            print("{0} worked with Jacobian {1}".format(method,str(jacobian)))
        except Exception as e:
            print("{0} FAILED with Jacobian {1} because {2}".format(method,str(jacobian), str(e)))
```
and got:
```
COBYLA worked with Jacobian None
COBYLA worked with Jacobian <function j at 0x0000019066F7C6A8>
COBYLA worked with Jacobian 2-point
Powell worked with Jacobian None
Powell worked with Jacobian <function j at 0x0000019066F7C6A8>
Powell worked with Jacobian 2-point
nelder-mead worked with Jacobian None
nelder-mead worked with Jacobian <function j at 0x0000019066F7C6A8>
nelder-mead worked with Jacobian 2-point
L-BFGS-B worked with Jacobian None
L-BFGS-B worked with Jacobian <function j at 0x0000019066F7C6A8>
L-BFGS-B worked with Jacobian 2-point
BFGS worked with Jacobian None
BFGS worked with Jacobian <function j at 0x0000019066F7C6A8>
BFGS worked with Jacobian 2-point
CG worked with Jacobian None
CG worked with Jacobian <function j at 0x0000019066F7C6A8>
CG worked with Jacobian 2-point
TNC worked with Jacobian None
TNC worked with Jacobian <function j at 0x0000019066F7C6A8>
TNC worked with Jacobian 2-point
SLSQP worked with Jacobian None
SLSQP worked with Jacobian <function j at 0x0000019066F7C6A8>
SLSQP worked with Jacobian 2-point
Newton-CG FAILED with Jacobian None because Jacobian is required for Newton-CG method
Newton-CG worked with Jacobian <function j at 0x0000019066F7C6A8>
Newton-CG FAILED with Jacobian 2-point because Jacobian is required for Newton-CG method
dogleg FAILED with Jacobian None because Jacobian is required for dogleg minimization
dogleg FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Hessian is required for dogleg minimization
dogleg FAILED with Jacobian 2-point because Jacobian is required for dogleg minimization
trust-ncg FAILED with Jacobian None because Jacobian is required for Newton-CG trust-region minimization
trust-ncg FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Either the Hessian or the Hessian-vector product is required for Newton-CG trust-region minimization
trust-ncg FAILED with Jacobian 2-point because Jacobian is required for Newton-CG trust-region minimization
trust-exact FAILED with Jacobian None because Jacobian is required for trust region exact minimization.
trust-exact FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Hessian matrix is required for trust region exact minimization.
trust-exact FAILED with Jacobian 2-point because Jacobian is required for trust region exact minimization.
trust-krylov FAILED with Jacobian None because ('Jacobian is required for trust region ', 'exact minimization.')
trust-krylov FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Either the Hessian or the Hessian-vector product is required for Krylov trust-region minimization
trust-krylov FAILED with Jacobian 2-point because ('Jacobian is required for trust region ', 'exact minimization.')
trust-constr worked with Jacobian None
trust-constr worked with Jacobian <function j at 0x0000019066F7C6A8>
trust-constr worked with Jacobian 2-point
```
Looks like a few other methods require a Jacobian: `trust-krylov`, `trust-exact`, and `dogleg`.

**Comment by mdhaber** [comment]
There are several formatting issues in the `minimize` documentation.
https://14717-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.optimize.minimize.html#scipy.optimize.minimize

image 
(shouldn't have vertical gray bar such places - caused by incorrect indentation)

image
`x`, `args`, `jac` need double back-ticks to render as teletype in RST.



Do you want to fix those while you're in there, or save for separate PR?

**Comment by mdhaber** [comment]
Let me know if I should open a separate issue for all this.

File: scipy/optimize/_minimize.py
Comment (mdhaber) [inline scipy/optimize/_minimize.py]:
```suggestion
        the fixed parameters. Methods 'Newton-CG', 'dogleg', 'trust-ncg', 'trust-exact', and 
        'trust-krylov'  require that a callable
```

**Review by mdhaber — CHANGES_REQUESTED** [review]

File: scipy/optimize/_minimize.py
Comment (andyfaff) [reply scipy/optimize/_minimize.py]: done

**Comment by andyfaff** [comment]
@mdhaber I've updated `jac` to refer to the extra methods requiring a callable jacobian. I also added the backticks you requested.

> Looks like a few other methods require a Jacobian: trust-krylov, trust-exact, and dogleg. These should probably be noted here, too.

done

> x, args, jac need double back-ticks to render as teletype in RST.

done

> "L-BFGS-B worked with Jacobian 3-point" (several methods work with '3-point', 'cs', but are they actually using those methods to get Jacobian information?)

If you trawl through the code you'll see that they work, but they should [issue a warning](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L547) that they're just going to use 2-point forward differences anyway.

**Comment by andyfaff** [comment]
Don't merge this just yet, I'm not sure if there is a mistake in the `minimize` function.

**Comment by andyfaff** [comment]
Closed, including in omnibus #10673

**Pull request closed without merging**
